Fix historacle memory issues

jvperrin opened the pull request:
I changed the historacle script to only read approximately every 100th line to use less memory.
This should resolve [rt #3954](https://rt.ocf.berkeley.edu/Ticket/Display.html?id=3954).

I also changed the loaded scripts to be loaded using `https` instead of `http` for a bit more security and removed some trailing whitespace and unused variables.

Here's a [demo](https://ocf.io/~jvperrin/index.php) of what it looks like with my changes.

chriskuehl commented:
thanks! you can close rt#3954 when this is merged:
https://rt.ocf.berkeley.edu/Ticket/Display.html?id=3954

@nattofriends is probably the best person to review this. LGTM though.

nattofriends commented:
this is a terrible approximation of sampling
i would rather it be done properly but chances of me caring enough to fix it myself are low

jvperrin commented:
Yeah, I'm definitely not an expert in PHP or anything, but I thought that just taking approximately every 100th point would work well enough for showing this graph. What would you recommend for doing it properly?

chriskuehl commented:
I'm guessing you could still read line-by-line and just only add every hundredth line to the array (I assume the array is where all the memory is being used -- that and the exporting as JSON).

Skipping a bunch of bytes is a little odd but probably fine in this case (each line is ~same number of bytes, so it's skipping a roughly consistent number of lines at least).

Not sure how to read line-by-line in PHP without loading the entire file into memory, but I assume it's possible...

nattofriends commented:
If you can muster the effort, reduce the memory consumption properly by fgetsing as now, and fseeking as appropriate based on the length of the line actually read.

If the file read is causing the memory consumption issue, stop here.
If the array storage is causing the memory consumption issue, perform your sampling in-loop by dropping read lines, instead of shooting for 2kb in that direction and calling it a day.

jvperrin commented:
The whole file was read into memory last time as an array, where the array was causing the memory issue I believe. I would have thought just skipping between lines by seeking would perform better than reading every line, and since the lines are roughly equal in length and contents, why would reading them help? The fseeking _is_ currently based on the length of the line read, since all the lines are approximately 20 bytes, so 2KB is approximately 100 lines, so I'm a bit confused about what you would like to change? Is it just to not hardcode the 2KB and instead read a sample line to determine the proper length to seek?

nattofriends commented:
What I was saying was to read all the lines using fgets, and then drop the lines using a counter or similar. This would let you have more control over how often you drop lines. If you don't want to do that, well there's nothing I can do about that.

chriskuehl commented:
@jvperrin up to you if you want to change it.

My idea is that long term labstats will just become a library (or merged into ocflib for simplicity) and we move the web code into ocfweb and get rid of the PHP entirely. So I probably wouldn't spend too much time on it.

jvperrin commented:
Well, I benchmarked both strategies, and reading and discarding unused lines actually uses less memory and (in my very limited tests) runs faster:

```
With skipping 2KB:

Memory used: 1668936 bytes
Time taken: 0.16963900566101 seconds
Time taken: 0.012646913528442
Time taken: 0.011445999145508
Time taken: 0.097899436950684

With discarding lines instead:

Memory used: 1505688 bytes
Time taken: 0.11171889305115 seconds
Time taken: 0.13188099861145
Time taken: 0.094919204711914
Time taken: 0.074958801269531
```

@nattofriends: You've got a good point that discarding lines allows for much finer control over what data is skipped, and it even performs better, which I didn't expect. I will change it to your method. Thanks for the feedback!

nattofriends commented:
Nice :chart_with_downwards_trend:

chriskuehl commented:
thank you github for `:chart_with_downwards_trend:`

nattofriends commented:
It showed up when I attempted to type `:metrics:`

chriskuehl commented:
bump, is this good to merge?

jvperrin commented:
I'm still slowly working on it, sorry for the wait. It should be good to go soon.

jvperrin committed:
Fix historacle memory issues

Select every 100th line of the printer csv files to reduce memory usage
Resolve rt #3954

jvperrin commented:
Ok, I've updated my changes to sample every 100th line and discard the other lines instead of seeking by 2KB chunks. I've updated my demo [here](https://ocf.io/~jvperrin/index.php) too, but there's not really any visible changes there, it's just the same graph essentially. Any suggestions?

nattofriends commented:
Looks good otherwise.

jvperrin commented:
Thanks!

jvperrin committed:
Use concise printer names and remove unused code

jvperrin commented:
Should I squash these two commits or no?

nattofriends commented:
I don't believing in squashing. Up to you.

Edit: Oh hey, I can press the merge button myself!
